### ISSUE TITLE
ngx.shared.DICT.lpush (and rpush) do not allow expiry time to be set

### ISSUE THREAD

**Issue by alexdowad** [openresty/lua-nginx-module#2123]
The documentation states that the function signature for `ngx.shared.DICT.lpush` is as follows:

**syntax:** *length, err = ngx.shared.DICT:lpush(key, value)*

It seems that there is no way to set the expiry time of a list which is added to a shared dictionary via `lpush` and `rpush`. Other API functions such as `ngx.shared.DICT.incr` accept an `init_ttl` argument, which specifies the expiry time of a key initialized by `incr`.

Since `lpush` and `rpush` do not set any expiry time for newly initialized lists, and do not remove unexpired items from memory, if I use `lpush` or `rpush` to populate a cache of items which are organized in lists, it seems that once the cache fills up, it will always be full and no new items will be saved until Nginx is restarted.

Am I missing something here?

If not, and if this feature is desired, I would be happy to send a PR adding the feature. (I would just need to know what the API should look like.)

**Comment by attenuation** [comment]
I think you are not understand https://github.com/openresty/lua-nginx-module#ngxshareddictlpush this API

`Inserts the specified (numerical or string) value at the head of the list named key in the shm-based dictionary ngx.shared.DICT. Returns the number of elements in the list after the push operation.`

This is mean this list which you insert is a shdict value (key is `key` args), it expire with this whole shdict value(this whose list), That is openresty shdcit design.

**Comment by alexdowad** [comment]
> `it will always be full and no new items` No, it will phasing out some shdict value to alloc memory for new value when the shared memory is full.

@attenuation, if that is the case, then the OpenResty documentation is wrong.

The documentation states (of `lpush`): 'It never overrides the (least recently used) unexpired items in the store when running out of storage in the shared memory zone. In this case, it will immediately return `nil` and the string "no memory".'

I am busy right now, but will post a link to the code which initializes a new shdict list a bit later. I believe it sets `expired` to zero, which means "never expire".

**Comment by alexdowad** [comment]
I think this is the relevant bit:

https://github.com/openresty/lua-nginx-module/blob/6e3b1c4fde47039f0ce2ee75fc678dab27944667/src/ngx_http_lua_shdict.c#L891-L904

I think that setting `sd->expires = 0` means that the new `SHDICT_TLIST` value inserted in the shdict will never automatically expire. Is that correct? Thanks in advance.

**Comment by attenuation** [comment]
Yes, but `sd->expires = 0` may still be deleted by LRU. https://github.com/openresty/lua-nginx-module/blob/6e3b1c4fde47039f0ce2ee75fc678dab27944667/src/ngx_http_lua_shdict.c#L1516

**Comment by alexdowad** [comment]
> Yes, but `sd->expires = 0` may still be deleted by LRU.
> 
> https://github.com/openresty/lua-nginx-module/blob/6e3b1c4fde47039f0ce2ee75fc678dab27944667/src/ngx_http_lua_shdict.c#L1516

Wow, this is interesting!

So I guess the documentation is incorrect? It says that `lpush` never overrides unexpired LRU items.
